File: selenium-csharp/Kestrel.Automation/Pages/CartPage.cs
namespace Kestrel.Automation.Pages;

public class CartPage
{
    private readonly IWebDriver _driver;
    private readonly Func<string, IWebElement> _el;

    public CartPage(IWebDriver driver, Func<string, IWebElement> el)
    {
        _driver = driver;
        _el = el;
    }

    public CartPage Open()
    {
        _driver.Navigate().GoToUrl($"{Config.BaseUrl}/cart.html");
        return this;
    }

    public string Subtotal => _el("summary-subtotal").Text;

    public IList<IWebElement> Rows() => _driver.FindElements(ByTestId.Id("cart-row"));

    public void ContinueToCheckout() => _el("checkout-link").Click();
}

File: playwright/pages/CartPage.ts
import { type Page, type Locator, expect } from "@playwright/test";

export class CartPage {
  readonly empty: Locator;
  readonly rows: Locator;
  readonly checkout: Locator;
  readonly subtotal: Locator;

  constructor(private readonly page: Page) {
    this.empty = page.getByTestId("empty-cart");
    this.rows = page.getByTestId("cart-row");
    this.checkout = page.getByTestId("checkout-link");
    this.subtotal = page.getByTestId("summary-subtotal");
  }

  async goto() {
    await this.page.goto("/cart.html");
  }

  async expectItem(name: string) {
    await expect(this.rows.filter({ hasText: name })).toBeVisible();
  }
}
